src/utils.ts

```diff
@@ -0,0 +1,84 @@
+/* eslint-disable no-template-curly-in-string, @typescript-eslint/no-explicit-any */
+ 
+import { homedir } from 'node:os'
+import path from 'node:path'
+import process from 'node:process'
+
+import vscode, { Uri } from 'vscode'
+
+/**
+ * code migrated from: https://github.com/tjx666/open-in-external-app/blob/master/src/parseVariables.ts
+ * 
+ * @see https://code.visualstudio.com/docs/editor/variables-reference
+ */
+export async function parseVariables(str: string, activeFile: Uri) {
+    const replacement: Map<string | RegExp, string | ((substring: string, ...args: any[]) => string) | undefined> =
+        new Map([
+            ['${userHome}', homedir()],
+            ['${pathSeparator}', path.sep],
+        ])
+
+    const { workspaceFolders } = vscode.workspace
+    if (workspaceFolders?.length) {
+        const workspaceFolder = workspaceFolders[0]!
+        replacement.set('${workspaceFolder}', workspaceFolder.uri.fsPath)
+        replacement.set('${workspaceFolderBasename}', workspaceFolder.name)
+    }
+
+    const { activeTextEditor } = vscode.window
+    const absoluteFilePath = activeFile.fsPath
+    replacement.set('${file}', absoluteFilePath)
+
+    const activeWorkspace = activeFile ? vscode.workspace.getWorkspaceFolder(activeFile) : undefined
+    replacement.set('${fileWorkspaceFolder}', activeWorkspace?.uri.fsPath)
+    replacement.set('${cwd}', activeWorkspace?.uri.fsPath)
+
+    let relativeFilePath: string | undefined
+    if (absoluteFilePath && activeWorkspace) {
+        relativeFilePath = absoluteFilePath.replace(activeWorkspace.uri.fsPath, '').slice(path.sep.length)
+        replacement.set('${relativeFile}', relativeFilePath)
+    }
+
+    if (relativeFilePath) 
+        replacement.set('${relativeFileDirname}', relativeFilePath.slice(0, relativeFilePath.lastIndexOf(path.sep)))
+
+    if (absoluteFilePath) {
+        const parsedPath = path.parse(absoluteFilePath)
+        replacement.set('${fileBasename}', parsedPath.base)
+        replacement.set('${fileBasenameNoExtension}', parsedPath.name)
+        replacement.set('${fileExtname}', parsedPath.ext)
+        replacement.set('${fileDirname}', parsedPath.dir)
+    }
+
+    if (activeTextEditor) {
+        replacement.set('${lineNumber}', String(activeTextEditor.selection.start.line + 1))
+
+        const cursorPosition = activeTextEditor.selection.active
+        replacement.set('${cursorLineNumber}', String(cursorPosition.line + 1))
+        replacement.set('${cursorColumnNumber}', String(cursorPosition.character + 1))
+
+        replacement.set(
+            '${selectedText}',
+            activeTextEditor.document.getText(
+                new vscode.Range(activeTextEditor.selection.start, activeTextEditor.selection.end),
```

Comment:
fyi, `activeTextEditor.selection` can be used here directly as its type compatible with Range "shape" [ts docs explanation](https://www.typescriptlang.org/docs/handbook/typescript-in-5-minutes.html#structural-type-system)